### ISSUE TITLE
Secondary clustering failing

### ISSUE THREAD

**Issue by shaman-narayanasamy** [MrOlm/drep#7]
Hi!

I am trying to cluster a large number of population-level genomes (5449 genomes after length filtering of dRep).

It seemed to have ran just fine until the secondary clustering step. I got the error below:
```
Step 1. Parse Arguments
Step 2. Perform MASH (primary) clustering
2a. Run pair-wise MASH clustering
2b. Cluster pair-wise MASH clustering
109 primary clusters made
Step 3. Perform secondary clustering
Running 143000 ANIn comparisons- should take ~ 1179.8 min
Traceback (most recent call last):
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/env3.4.3/bin/dRep", line 26, in <module>
    controller.parseArguments(args)
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-p
ackages/drep/controller.py", line 136, in parseArguments
    self.dereplicate_wf_operation(**vars(args))
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/drep/controller.py", line 80, in dereplicate_wf_operation
    drep.d_workflows.dereplicate_wrapper(kwargs['work_directory'],**kwargs)
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/drep/d_workflows.py", line 36, in dereplicate_wrapper
    drep.d_cluster.d_cluster_wrapper(wd, **kwargs)
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/drep/d_cluster.py", line 287, in d_cluster_wrapper
    Cdb, Mdb, Ndb = cluster_genomes(Bdb, data_folder, **kwargs)
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/drep/d_cluster.py", line 119, in cluster_genomes
    **kwargs)
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/drep/d_cluster.py", line 153, in run_secondary_clustering
    ndb = compare_genomes(bdb, algorithm, data_folder, **kwargs)
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/drep/d_cluster.py", line 866, in compare_genomes
    df = run_pairwise_ANIn(genome_list, working_data_folder, **kwargs)
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/drep/d_cluster.py", line 913, in run_pairwise_ANIn
    for x,y in zip(genomes,genomes)}
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/drep/d_cluster.py", line 913, in <dictcomp>
    for x,y in zip(genomes,genomes)}
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/drep/__init__.py", line 64, in fasta_length
    for seq_record in SeqIO.parse(fasta, "fasta"):
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/Bio/SeqIO/__init__.py", line 600, in parse
    for r in i:
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/Bio/SeqIO/FastaIO.py", line 122, in FastaIterator
    for title, sequence in SimpleFastaParser(handle):
  File "/mnt/nfs/projects/ecosystem_biology/local_tools/pyenv/versions/3.4.3/envs/env3.4.3/lib/python3.4/site-packages/Bio/SeqIO/FastaIO.py", line 62, in SimpleFastaParser
    line = handle.readline()
BrokenPipeError: [Errno 108] Cannot send after transport endpoint shutdown
```

Note that dRep worked fine for us when applying it onto smaller data sets. Any idea why this happened? I'm looking forward to your reply.

Best regards,
Shaman

**Comment by MrOlm** [comment]
Hi Shaman,

Glad that dRep is working on smaller datasets, and I'm sorry that this error is happening on your larger dataset.

The error is happening when the program is simply calculating the length of the input .fasta files, which is a weird place for it to crash. The error seems to be with BioPython, and from googling it might have to do with the computer system you're running the program on. If it does have to do with your system architecture I won't be able to help troubleshoot very much. It's especially strange because a identical piece of code is used to filter the genomes by length in the beginning of the program, so it's strange the program didn't crash there.

I pushed a special branch of dRep for you to download to help us troubleshoot this. The main difference is that it logs which genome it's trying to calculating the length of before it does so, so we can see which genome is causing the program to crash, and go from there. It also calculates the length in a slightly different way that may fix the problem altogether.

The new branch is called "shaman_fix". To download it, run:

`git clone -b shaman_fix git@github.com:MrOlm/drep.git`

Then install it with pip however you did previously. For example:

```
cd drep
pip install .
```

If successful, when you run the help you should see the following:

```
$ dRep -h

                ...::: dRep v0.5.4_SHAMANFIX :::...

  Choose one of the operations below for more detailed help.
  Example: dRep dereplicate_wf -h
```

Once you have this set up, go ahead and run the command again. If you run the exact same command (with the same directory and everything), the program should be able to find the past secondary clustering runs and not have to run them again. You may have to include `-o` and you may have to not include the list of genomes with `-g`, as it may have already stored the location of all of them (the program will tell you if either of these is the case).

Once this new run is complete, we can check the log (located in `log/logger.log`) to see which genome is causing the problem.

Please let me know if you have any trouble doing any of this, and please let me know how it turns out so that I can advice others who run into this problem.

Best,
-Matt

**Comment by shaman-narayanasamy** [comment]
Hi Matt,

Thanks very much for the quick response! It works now with the new fix. 

Indeed it is strange. As I mentioned, the version without the fix works on smaller data sets. 

Thanks again for the fix, I guess you will incorporate it into the master branch and in the next version. I will look out for that and for dRep to be published as well :)

Best,
Shaman

**Comment by MrOlm** [comment]
Hi Shaman,

Glad the fix worked! I'm still not entirely sure why what I did fixed it, but I'm glad it did

The fix has been merged into master (version 0.5.5) and that branch has been removed.

Thanks for helping me troubleshoot this, and let me know if you run into any more problems

-Matt
